File: Min_Element_Rotated_Sorted_Array.java
// 153. Find Minimum in Rotated Sorted Array - https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/
// Time Complexity : O(logn)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution 
{
    public int findMin(int[] nums) 
    {
        if (nums.length == 1) // one element
        {
            return nums[0];
        }
        int left = 0, right = nums.length - 1;
        if (nums[right] > nums[0])     
        {
            return nums[0];
        }
      
        while (right >= left) 
        {
            int mid = left + (right - left) / 2;
            if (nums[mid] > nums[mid + 1]) 
            {
                return nums[mid + 1];
            }
  
            if (nums[mid - 1] > nums[mid]) 
            {
                return nums[mid];
            }
            if (nums[mid] > nums[0]) 
            {
                left = mid + 1;
            } 
            else 
            {
                right = mid - 1;
            }
        }
        return -1;
    }
}

File: MinElementRotatedSortedArray.kt
class Solution {
    fun findMin(nums: IntArray): Int {

        var left = 0
        var right = nums.size - 1

        while (left <= right) {

            val mid = left + (right - left) / 2
            print(mid)

            if ( ( mid == 0 || nums[mid - 1] > nums[mid]) && (mid == nums.size - 1 || nums[mid + 1] > nums[mid]))
            {return nums[mid]}

            if (nums[mid] < nums[right]) {
                right = mid -1
            } else {
                left = mid + 1
            }

        }

        return -1
    }
}

